src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,16 +33,28 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        internal static T EnsureInitializedDisposable<T>([AllowNull] ref T? field, Func<T> initializer) where T : class, IDisposable
+            => Volatile.Read(ref field) ?? EnsureInitializedDisposableCore(ref field, initializer);
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+        private static T EnsureInitializedDisposableCore<T>([AllowNull] ref T? field, Func<T> initializer) where T : class, IDisposable
+        {
+            T value = initializer();
+
+            if (Interlocked.CompareExchange(ref field, value, null) != null)
+            {
+                value.Dispose();
+            }
+
+            return field;
+        }
+
+        public static TextReader In => EnsureInitializedDisposable(ref s_in, () => ConsolePal.GetOrCreateReader());
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                return LazyInitializer.EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
```

Comment:
Is there now an issue between threads concurrently accessing InputEncoding's getter and setter?  The setter is using a lock that the getter is no longer using, and the setter is doing multiple writes to distinct fields.  Seems like this fields could now get out of sync if the getter and setter are used concurrently?

Reply:
Ah yes, we need to continue using the lock here.

Reply:
Looking a bit closer at the code, I don't see an issue since `s_inputEncoding` will end up with the proper `Encoding` after a set, and the getter cannot replace a set encoding.

```cs
            get
            {
                return LazyInitializer.EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
            }
            set
            {
                CheckNonNull(value, nameof(value));
                lock (InternalSyncObject)
                {
                    // Set the terminal console encoding.
                    ConsolePal.SetConsoleInputEncoding(value);

                    Volatile.Write(ref s_inputEncoding, (Encoding)value.Clone());

                    // We need to reinitialize 'Console.In' in the next call to s_in
                    // This will discard the current StreamReader, potentially
                    // losing buffered data.
                    Volatile.Write(ref s_in, null);
                }
            }
```
Do you also think it's ok?

Reply:
For `OutputEncoding`, the `Volatile.Write` to `s_out` and `s_error` should also be moved after the `s_outputEncoding` write.